### PR TITLE
[output] Elasticsearch output plugin

### PR DESCRIPTION
**Pull request by ScienJus**
this draft pull request is the implementation of #127 , and is still work in progress.

After code review and running a case in the real world, I will add the document.😿

### CODE REVIEW

**Comment by Ryan-Git** [comment]
LGTM

File: pkg/outputs/elasticsearch/elasticsearch.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+func (output *ElasticsearchOutput) Execute(msgs []*core.Msg) error {
+	var reqs []elastic.BulkableRequest
+	for _, msg := range msgs {
+		// only support dml message
+		if msg.DmlMsg == nil {
+			continue
+		}
+		// must have a primary key
+		if len(msg.DmlMsg.Pks) == 0 {
+			continue
```
Comment (ming535) [inline pkg/outputs/elasticsearch/elasticsearch.go]: If this plugin does not support source table without primary key, we should return error with detailed information here, and tell the user that we do not support it in the error message. The scheduler will panic and stop the process.

File: pkg/outputs/elasticsearch/elasticsearch.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+				// tags: [pipelineName, index, action(index/create/delete/update)].
+				// indices created in 6.x only allow a single-type per index, so we don't need the type as a tag.
+				metrics.OutputCounter.WithLabelValues(output.pipelineName, result.Index, action, "", "").Add(1)
+			} else if result.Status == http.StatusNotFound && action == "delete" {
+				// delete but not found, just ignore it.
+				metrics.OutputCounter.WithLabelValues(output.pipelineName, result.Index, action, "", "").Add(1)
+				continue
+			} else if result.Status == http.StatusTooManyRequests {
+				// when the server returns 429, it must be that all requests have failed.
+				time.Sleep(5 * time.Second)
```
Comment (ming535) [inline pkg/outputs/elasticsearch/elasticsearch.go]: The scheduler will retry if `Execute` returns error in `batch_table_scheduler.go`

File: pkg/outputs/elasticsearch/elasticsearch.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+				return output.sendBulkRequests(reqs)
+			} else if output.config.IgnoreBadRequest {
```
Comment (ming535) [inline pkg/outputs/elasticsearch/elasticsearch.go]: can we combine the first three `if` cases together?

File: pkg/outputs/elasticsearch/elasticsearch.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+				// TODO do we need a metcris in prometheus for this?
```
Comment (ming535) [inline pkg/outputs/elasticsearch/elasticsearch.go]: may be we can just reuse `OutputCounter `

File: pkg/outputs/elasticsearch/elasticsearch.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+				log.Infof("[output_elasticsearch] Received and ignored an error from server, status: [%d], index: %s, details: %s.", result.Status, result.Index, marshalError(result.Error))
+			} else {
+				log.Panicf("[output_elasticsearch] Received an error from server, status: [%d], index: %s, details: %s.", result.Status, result.Index, marshalError(result.Error))
```
Comment (ming535) [inline pkg/outputs/elasticsearch/elasticsearch.go]: we can just return the error to scheduler with detailed error information. The scheduler will panic if multiple retries failed.

**Review by ming535 — CHANGES_REQUESTED** [review]

**Comment by ScienJus** [comment]
@ming-relax Thanks, please take a look again.

**Comment by ming535** [comment]
@ScienJus Let's add the document into this PR

**Review by ming535 — APPROVED** [review]

**Pull request merged**